### PR TITLE
test: Implement GSP-86 Add linker integration tests

### PR DESCRIPTION
**Pull request by abyss-w**
test: Implement GSP-86 Add linker integration tests

### CODE REVIEW

File: linker.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+)
+
+func TestLinker(t *testing.T, store types.Storager) {
+	Convey("Given a basic Storager", t, func() {
+		l, ok := store.(types.Linker)
+		So(ok, ShouldBeTrue)
+
+		Convey("When create a link object", func() {
+			path := uuid.New().String()
+			_ = store.Create(path)
```
Comment (JinnyYi) [inline linker.go]: `Create` is a local function without sending API/RPC call，so `path` is a non-existent resource for server here. Maybe we can call `Write` to upload an object or file here as the source?

Comment (abyss-w) [reply linker.go]: Got it!

File: linker.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+			target := uuid.New().String()
+			o, err := l.CreateLink(path, target)
+
+			defer func() {
+				err = store.Delete(target)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
```
Comment (JinnyYi) [inline linker.go]: `path` and `target` are not exist, I think there's no need to delete them.

File: linker.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+			Convey("The first returned error should not be nil", func() {
+				So(err, ShouldNotBeNil)
```
Comment (JinnyYi) [inline linker.go]: ~~Maybe we could check whether the error is `ErrObjectNotExist` explicitly?~~
Actually we can create symlink for files that don't exist.

File: linker.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+			Convey("The linkTarget of the object must be the same as the target", func() {
+				// The linkTarget must be the same as the target.
+				linkTarget, ok := o.GetLinkTarget()
+
+				So(ok, ShouldBeTrue)
+				So(linkTarget, ShouldEqual, target)
+			})
+		})
+
+		Convey("When create a link object to a not existing target", func() {
```
Comment (JinnyYi) [inline linker.go]: I think it's the same with `When create a link object`?

Maybe we need to add the following cases:
- When CreateLink to an existing target
~~- When Stat with ModeLink~~
~~- When List with ModeLink~~
- ...

File: linker.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
+package tests
+
+import (
+	"github.com/beyondstorage/go-storage/v4/types"
+	"github.com/google/uuid"
+	. "github.com/smartystreets/goconvey/convey"
+	"testing"
+)
```
Comment (JinnyYi) [inline linker.go]: Let's format the import:

```
import (
    <golang stdlib>

    <3rd lib>

    <self lib>
)
```

File: linker.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			Convey("The first returned error should not be nil", func() {
+				So(err, ShouldNotBeNil)
+			})
+
+			o, err = l.CreateLink(path, target)
```
Comment (JinnyYi) [inline linker.go]: Maybe there's no need to create link twice for this case?

File: linker.go (hunk 4 above)
Comment (JinnyYi) [reply linker.go]: Maybe we need to check the object mode after a successful`CreateLink`.

File: go.mod [hunk 7]
```diff
@@ -3,7 +3,7 @@ module github.com/beyondstorage/go-integration-test/v4
 go 1.15
 
 require (
-	github.com/beyondstorage/go-storage/v4 v4.3.2
+	github.com/beyondstorage/go-storage/v4 v4.3.3-0.20210720033749-df6a96af6045
```
Comment (Xuanwo) [inline go.mod]: ping @JinnyYi, Let's solve https://github.com/beyondstorage/go-storage/issues/682 so that we can use it here.

Comment (JinnyYi) [reply go.mod]: `go-storage` v4.4.0 have been released.

**Comment by JinnyYi** [comment]
I made some mistakes and tested symlink based on files in `fs` and objects in `oss`:

- When `target` is exist and `path` is not exist:
  - Both `fs` and `oss` can create symlink successfully and the returned symlink is readable.
- When `target` is not exist  and `path` is not exist:
  - Both `fs` and `oss` can create symlink successfully.
  - The target can be get from the symlink.
  - After we write to `target`(the target file or object), we can get the content from the symlink.
- When `target` is exist and `path` is also exist:
  - For `oss`, `PutSymlink` will succeed no mater `path` identifies an object or a symlink, like `overwrite`.
  - For `fs`, `os.Symlink` will get error no mater `path` identifies a file, a directory or a symlink.

- For `oss`, the `Symlink` object type can be get from `ListObject`, but seems can't get form `GetMetadata`.
- For `fs`, maybe we should check and delete the `path` if it is exist as describe in [GSP-134: Write Behavior Consistency](https://github.com/beyondstorage/go-storage/blob/master/docs/rfcs/134-write-behavior-consistency.md).

**Comment by Xuanwo** [comment]
Nice catch, maybe we need to add them in specs or update the rfc.

**Comment by Xuanwo** [comment]
Linker should also take #42 a look.

File: linker.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+			Convey("Read should get path object data without error", func() {
+				var buf bytes.Buffer
+				n, err := store.Read(path, &buf)
+
+				Convey("The error should be nil", func() {
+					So(err, ShouldBeNil)
+				})
+
+				Convey("The content should be match", func() {
+					So(n, ShouldNotBeNil)
```
Comment (JinnyYi) [inline linker.go:163]: `err` should not be nil here.

File: linker.go [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+				var buf bytes.Buffer
+				n, err := store.Read(path, &buf)
+
+				Convey("The error should be nil", func() {
+					So(err, ShouldBeNil)
+				})
+
+				Convey("The content should be match", func() {
+					// The content should match the secondTarget
+					So(n, ShouldNotBeNil)
```
Comment (JinnyYi) [inline linker.go:246]: ditto

File: linker.go [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+			})
+
+			Convey("The object should be nil", func() {
+				So(o, ShouldBeNil)
+			})
+		})
+	})
+}
+
+func TestLinkerWithVirtualLink(t *testing.T, store types.Storager) {
```
Comment (JinnyYi) [inline linker.go]: This function will be called by services that support `overwrite`, like `oss`. But actually `symlink` is native support for `oss`. So maybe the function name `TestLinkerWithVirtualLink` is not suitable?

Comment (abyss-w) [reply linker.go]: How about  using ` TestLinkerWithoutOverwrite` and ` TestLinkerWithOverWrite`  as function names?

**Comment by Xuanwo** [comment]
This PR lives so long. 

How about removing the parts we haven't agreed on yet and merge `TestLinker` into master first?

We can specify the corner case behavior later.

**Comment by abyss-w** [comment]
> How about removing the parts we haven't agreed on yet and merge `TestLinker` into master first?

OK

File: linker.go [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+			defer func() {
+				err = store.Delete(target)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			Convey("Read should get path object data without error", func() {
+				var buf bytes.Buffer
+				n, err := store.Read(path, &buf)
```
Comment (Xuanwo) [inline linker.go]: `LinkObject` is not readable, so we can't read from a link directly. Instead, we need to read from the link object's `LinkTarget`.

Maybe we can remove tests about `Read should get path object data without error`?

Comment (abyss-w) [reply linker.go]: Got it!

**Comment by Xuanwo** [comment]
@abyss-w Hi, we will squash all commits while merging, so don't bother to do a force push here.

**Comment by abyss-w** [comment]
> @abyss-w Hi, we will squash all commits while merging, so don't bother to do a force push here.

ok, got it.

**Comment by Xuanwo** [comment]
ping @JinnyYi to take another look.

**Review by Xuanwo — APPROVED** [review]

**Comment by Xuanwo** [comment]
Thank you for your patient and detailed work!

**Pull request merged**